packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -4157,3 +4164,244 @@ class _CopySelectionAction extends ContextAction<CopySelectionTextIntent> {
   @override
   bool get isActionEnabled => state._value.selection.isValid && !state._value.selection.isCollapsed;
 }
+
+/// A void function that takes a [TextEditingValue].
+@visibleForTesting
+typedef TextEditingValueCallback = void Function(TextEditingValue value);
+
+/// Provides undo/redo capabilities for text editing.
+///
+/// Listens to [controller] as a [ValueNotifier] and saves relevant values for
+/// undoing/redoing. The cadence at which values are saved is a best
+/// approximation of the native behaviors of a hardware keyboard on Flutter's
+/// desktop platforms, as there are subtle differences between each of these
+/// platforms.
+///
+/// Listens to keyboard undo/redo shortcuts and calls [onChanged] when a
+/// shortcut is triggered that would affect the state of the [controller].
+@visibleForTesting
+class TextEditingHistory extends StatefulWidget {
+  /// Creates an instance of [TextEditingHistory].
+  const TextEditingHistory({
+    Key? key,
+    required this.child,
+    required this.controller,
+    required this.onChanged,
+  }) : super(key: key);
+
+  /// The child widget of [TextEditingHistory].
+  final Widget child;
+
+  /// The [ValueNotifier] to save the state of over time.
+  final TextEditingController controller;
```

Comment:
This can be reduced to value notifier according to the doc comment.

not too much of an issue if this is a private class though

Reply:
Ah this is a comment that I forgot to update, I'll fix it.

Originally I tried to make this widget work with any type of ValueNotifier, but it turns out that there is a good amount of text-editing-specific logic that is needed to get undo/redo to work with EditableText.  For example, ignoring selection and composing updates.  I think it makes sense to just put that here since we have no use for a more generic undo/redo widget now.

Reply:
if we are not modifying the value in controller in this widget a ValueNotifier of TextEditingValue should be enough, no?

but since this class can be private, I think it is probably fine as is